language/stdlib/modules/association.move

```diff
@@ -109,6 +114,156 @@ module Association {
+        define spec_root_address(): address { 0xA550C18 }
+
+        // mirrors Move addr_is_association
+        define spec_addr_is_association(addr: address): bool {
+            exists<PrivilegedCapability<T>>(addr)
+            && global<PrivilegedCapability<T>>(addr).is_certified
+        }
+     }
+
+    // ## Management of privileges
+    // TODO: With grant/remove etc., separately specify that only root may grant or remove privileges
+
+    // ### Management of Root {} marker.
+
+    // The root_address is marked by a Root{} object that is stored at that place only.
+    spec module {
+        // enclose this in a function so we can easily repeat the same property in the invariant
+        // and as a post-condition to initialize (the base case).
+        // SPEC: Only the root address has a Root{} resource.
+        define root_invariant(): bool {
+            all(addresses(), |a| exists<Root>(a) ==> (a == spec_root_address()))
+        }
+    }
+    spec schema OnlyRootAddressHasRootPrivilege {
+        invariant root_invariant();
+    }
+    spec module {
+        // apply to all functions except "initialize" and functions that "initialize" calls
+        // before the invariant is established.
+        // Note: All the called functions OBVIOUSLY cannot affect the invariant.
+        // TODO: Try to find a better approach to this that does not require excepting functions.
+        apply OnlyRootAddressHasRootPrivilege to *<T>, *
+            except initialize, root_address, has_privilege, addr_is_association,
+            assert_addr_is_association, assert_sender_is_association;
+    }
+    spec fun initialize {
+        // "initialize" establishes the invariant, so it's a special case.
+        // Before initialize, no addresses have a Root{} resource.
+        // Afterwards, only Root {} has the root resource.
+        requires all(addresses(), |a| !exists<Root>(a));
+        ensures root_invariant();
+    }
+
+    // This post-condition to assert_sender_is_root is a sanity check that
+    // the Root {} invariant really works. It needs the invariant
+    // OnlyRootAddressHasRootPrivilege, because it assert_sender_is_root does not
+    // directly check that the sender == root_address(). Instead, it aborts if
+    // sender has root privilege (Root{}), and only the root_address has Root{}
+    // TODO: There is a style question about whether this should just check for presence of
+    // a Root privilege. I guess it's moot so long as OnlyRootAddressHasRootPrivilege holds.
+    spec fun assert_sender_is_root {
+        ensures sender() == spec_root_address();
+    }
+
+    // ### authority to set is_certified flag
+
+    // SPEC: only grant_* functions can set the is_certified flag in
+    // PrivilegedCapability<Privilege>.
+    // Other functions can create and remove the Privilege object, so we have to condition
+    // everything on its existence.
+    spec schema OnlyGrantCanCertify<Privilege> {
```

Comment:
This schema would be associated with function `assert_sender_is_root` in the generated doc, so you probably want to move it into the module stuff before any function, where the other apply is.